src/hotspot/share/opto/movenode.cpp

```diff
@@ -185,6 +191,65 @@ CMoveNode *CMoveNode::make(Node *c, Node *bol, Node *left, Node *right, const Ty
   }
 }
 
+// Try to identify min/max patterns in CMoves
+Node* CMoveNode::Ideal_minmax(PhaseGVN* phase, CMoveNode* cmove) {
+  // If we're post loop opts then don't attempt to match the min/max pattern, as this node might have been a
+  // MinL or MaxL that was already expanded during macro expansion.
+  if (phase->C->post_loop_opts_phase()) {
+    return nullptr;
+  }
+
+  // The BoolNode may have been idealized into a constant. If that's the case, then Identity should take care of it instead.
+  BoolNode* bol = cmove->in(CMoveNode::Condition)->isa_Bool();
+  if (bol == nullptr) {
+    return nullptr;
+  }
+
+  Node* cmp = bol->in(1);
+  int cmove_op = cmove->Opcode();
+  int cmp_op = cmp->Opcode();
+
+  // Ensure comparison is an integral type, and that the cmove is of the same type.
+  if ((cmp_op != Op_CmpI || cmove_op != Op_CMoveI) && (cmp_op != Op_CmpL || cmove_op != Op_CMoveL)) {
```

Comment:
What if we combine a `CmpI` with a `CMoveL`?
Or maybe there is some strange way to use a `Float.floatToIntBits` and combine a `CmpI` with a `CMoveF`?

Ah, wait. I think it is correct. It's just difficult to read these "inverted" formulas.
I suggest you rewrite it to be:
`! (both-int or both-long)`

Reply:
Ah yep, I had originally wrote it that way but I factored out the `!`. I agree that it would be cleaner that way, though.